x/gasless/keeper/fee_helper.go

```diff
@@ -0,0 +1,200 @@
+		return originalFeePayer
+	}
+
+	// only one fee coin is supported, tx containing multiple coins as fees are not allowed.
+	if len(fees) != 1 {
+		k.EmitFeeConsumptionEvent(ctx, originalFeePayer, []uint64{}, []error{sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "asked fee != 1")}, 0)
+		return originalFeePayer
+	}
+
+	fee := fees[0]
+
+	usageIdentifier := k.ExtractUsageIdentifierFromTx(ctx, sdkTx)
+
+	// check if there are any gas tanks for given usageIdentifier
+	// if there is no gas tank for the given identifier, fee source will be original feePayer
+	usageIdentifierToGasTankIds, found := k.GetUsageIdentifierToGasTankIds(ctx, usageIdentifier)
+	if !found {
+		k.EmitFeeConsumptionEvent(ctx, originalFeePayer, []uint64{}, []error{sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "no gas tanks found")}, 0)
+		return originalFeePayer
+	}
+
+	tempConsumer, found := k.GetGasConsumer(ctx, originalFeePayer)
+	if !found {
+		tempConsumer = types.NewGasConsumer(originalFeePayer)
+	}
+
+	var (
+		failedGtids      []uint64
+		failedGtidErrors []error
+		gasTank          types.GasTank
+		isValid          bool
+		err              error
+	)
+
+	gasTankIds := usageIdentifierToGasTankIds.GasTankIds
+	for _, gtid := range gasTankIds {
+		gasTank, isValid, err = k.CanGasTankBeUsedAsSource(ctx, gtid, tempConsumer, fee)
+		if isValid {
+			break
+		}
+		failedGtidErrors = append(failedGtidErrors, err)
+		failedGtids = append(failedGtids, gtid)
+	}
+
+	if !isValid {
+		k.EmitFeeConsumptionEvent(ctx, originalFeePayer, failedGtids, failedGtidErrors, 0)
+		return originalFeePayer
+	}
+
+	// update the consumption and usage details of the consumer
+	gasConsumer, consumptionIndex := k.GetUpdatedGasConsumerAndConsumptionIndex(ctx, originalFeePayer, gasTank, fee.Amount)
+
+	existingUsage := gasConsumer.Consumptions[consumptionIndex].Usage
+
+	usageIdentifierFound := false
+	usageIdentifierIndex := 0
+
+	for index, usageDetail := range existingUsage {
+		if usageDetail.UsageIdentifier == usageIdentifier {
+			usageIdentifierFound = true
+			usageIdentifierIndex = index
+		}
+	}
+
+	if !usageIdentifierFound {
+		existingUsage = append(existingUsage, &types.Usage{
+			UsageIdentifier: usageIdentifier,
+			Details:         []*types.Detail{},
+		})
+		usageIdentifierIndex = len(existingUsage) - 1
+	}
+
+	existingUsage[usageIdentifierIndex].Details = append(existingUsage[usageIdentifierIndex].Details, &types.Detail{
+		Timestamp:   ctx.BlockTime(),
+		GasConsumed: fee.Amount,
+	})
+
+	// assign the updated-existingUsage usage and set it to the store
+	gasConsumer.Consumptions[consumptionIndex].Usage = existingUsage
+	k.SetGasConsumer(ctx, gasConsumer)
+
+	// shift the used gas tank at the end of all tanks, so that a different gas tank can be picked
+	// in next cycle if there exists any.
+	usageIdentifierToGasTankIds.GasTankIds = utils.ShiftValueToEnd(usageIdentifierToGasTankIds.GasTankIds, gasTank.Id)
+	k.SetUsageIdentifierToGasTankIds(ctx, usageIdentifierToGasTankIds)
```

Comment:
please document this in the spec too
I understand this is to spread out usage?
It's not a given that this is the optimal way for all use cases
Some uses might prefer simple first come first served etc